src/System.Private.ServiceModel/src/System/ServiceModel/Channels/StreamSecurityUpgradeInitiatorBase.cs

```diff
@@ -99,33 +75,48 @@ public override Stream InitiateUpgrade(Stream stream)
             return result;
         }
 
-        internal override void EndOpen(IAsyncResult result)
+        internal override async Task<Stream> InitiateUpgradeAsync(Stream stream)
         {
-            base.EndOpen(result);
-            _isOpen = true;
+            if (stream == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("stream");
+            }
+
+            if (!_isOpen)
+            {
+                 Open(TimeSpan.Zero);
+            }
+
+            var remoteSecurityWrapper = new OutWrapper<SecurityMessageProperty>();
+            Stream result = await OnInitiateUpgradeAsync(stream, remoteSecurityWrapper);
+            _remoteSecurity = remoteSecurityWrapper;
+            _securityUpgraded = true;
+            return result;
         }
 
         internal override void Open(TimeSpan timeout)
         {
-            base.Open(timeout);
             _isOpen = true;
         }
 
-        internal override void EndClose(IAsyncResult result)
+        internal override Task OpenAsync(TimeSpan timeout)
         {
-            base.EndClose(result);
-            _isOpen = false;
+            _isOpen = true;
```

Comment:
Call the synchronous Open method instead of duplicating the logic. That way any code changes happen in one place. The JIT compiler should inline it so there won't be any impact at runtime.